### DISCUSSION TITLE
Get all hard linked rows with check access

### DISCUSSION THREAD

**Discussion by klaydev** [sqlalchemy/sqlalchemy#10983] [Usage Questions]
Hi. I have a function that gets the Product Spec from the database with access verification. Access verification is performed using the following algorithm: get a Product Spec, get a Product linked to it, get a Shop linked to a Product, compare Shop.user_id and User.id (models are shown below)
The first problem: in order to make a check in this way when receiving one ProductSpec, you need to get a row from the Many-To-Many table, get the product from there, get the store from the product, compare the IDs, return None if the IDs do not match. If the IDs match, then you need to get the Product Spec.
This approach is bad because too many unnecessary requests are being made. 
This leads to the first question, how can this logic be optimized?
The second problem: if you need to get all ProductSpec, in this case you need to load all the rows from the Many-To-Many table, then iterate through each and get each linked Product, get each linked Shop and only then compare the identifiers. After that, you need to get all the ProductSpec and return it. In this case, the problems with extra requests are enormous, because with a million records, you need to make at least 5 million requests to get all productspecs, which is not at all optimal.
This leads to the second question, how to optimize this logic?

I will also attach a link to the answer on the topic, which will clarify what is happening
https://github.com/sqlalchemy/sqlalchemy/discussions/10953
```py
async def get_product_spec_from_database(id: UUID, user: User, session: AsyncSession) -> ProductSpecBase | None:
    row_spec = await session.scalar(select(ProductSpec)
                                .where(foreign(func.coalesce(
                                                  *(column for column in inspect(ProductSpec).columns if
                                                    "spec" in column.name)
                                              )) == id)
                                )
    product = await session.get(Product, row_spec.product_id)
    if product.shop.user_id != user.id:
        return None

    spec = await session.scalar(select(ProductSpecBase)
                                .where(ProductSpecBase.id == foreign(func.coalesce(
                                                  *(column for column in inspect(ProductSpec).columns if
                                                    "spec" in column.name)
                                              ))
                                       )
                                .join(ProductSpec, foreign(func.coalesce(
                                                  *(column for column in inspect(ProductSpec).columns if
                                                    "spec" in column.name)
                                              )) == id)
                                )

    await session.refresh(spec)
    return spec
```
```py
# models
class ProductSpec(Base):
    __tablename__ = "product_spec"
    __table_args__ = (
        CheckConstraint(
            "(kit_spec_id IS NOT NULL)::int +" +
            "(case_spec_id IS NOT NULL)::int +" +
            "(privilege_spec_id IS NOT NULL)::int = 1"
        ),
    )

    kit_spec_id: Mapped[UUID | None] = mapped_column(ForeignKey("product_spec_kit_products.id"))
    case_spec_id: Mapped[UUID | None] = mapped_column(ForeignKey("product_spec_case_products.id"))
    privilege_spec_id: Mapped[UUID | None] = mapped_column(ForeignKey("product_spec_privilege.id"))
    product_id: Mapped[UUID] = mapped_column(ForeignKey("product.id"), index=True, unique=True, primary_key=True)


class ProductSpecBase(AbstractConcreteBase, Base):
    id: Mapped[uuidpk]


class Product(Base):
    __tablename__ = "product"

    ...
    shop_id: Mapped[UUID] = mapped_column(ForeignKey("shop.id"), index=True)
    shop = relationship("src.consumers.models.shop.Shop", lazy="joined")

class Shop(Base):
    __tablename__ = "shop"

    ...
    user_id: Mapped[UUID] = mapped_column(ForeignKey("user.id"), index=True)
    user = relationship("src.auth.models.User", lazy="joined")
```

**Comment by zzzeek** [comment]
> This approach is bad because too many unnecessary requests are being made.

getting data from multiple tables with fewer requests, use a join?  since you have a concrete mapping, also need to use UNION.  it looks like you have joins there.  is there a SQL statement that would solve the problem that we can illustrate how to invoke here with the ORM model you have?   let's think in terms of the tables.    Would things be easier if the subclasses of ProductSpecBase all came from a common table instead?


> The second problem: if you need to get all ProductSpec, in this case you need to load all the rows from the Many-To-Many table, then iterate through each and get each linked Product, get each linked Shop and only then compare the identifiers. After that, you need to get all the ProductSpec and return it. In this case, the problems with extra requests are enormous, because with a million records, you need to make at least 5 million requests to get all productspecs, which is not at all optimal.
This leads to the second question, how to optimize this logic?

again are we looking to solve this in SQL first ?   we want to think in terms of SQL at all times.    I can SELECT from all these tables at once, can we work out that query?
